### ISSUE TITLE
Route Bound Parameters incorrect when using "::"

### ISSUE THREAD

**Issue by paratiger** [rails/rails#53397]
### Steps to reproduce

Creating a route that contains "::" in bound parameters no longer works in Rails 8rc1

An example would would be "articles(/sort::sort)" which in previous versions would create a route such as "articles/sort:title" with params[:sort] returning 'title'

Below is a script showing this issue

```ruby
# frozen_string_literal: true
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "~> 8.0.0.rc1"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"
end

require "action_controller/railtie"

class TestApp < Rails::Application

  config.root = __dir__
  config.hosts << "example.org"
  config.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/(sort::sort)" => "test#index"
  end

end

class TestController < ActionController::Base

  include Rails.application.routes.url_helpers

  def index
    puts params
    render plain: params[:sort]
  end

end

require "minitest/autorun"
require "rack/test"

class BugTest < ActiveSupport::TestCase

  include Rack::Test::Methods

  def test_returns_success
    get "/sort:sample"
    assert last_response.body == 'sample'
  end

  private

  def app
    Rails.application
  end

end
```

### Expected behavior

We would expect params[:sort] to return 'sample' with the output from params returning `{"controller"=>"test", "action"=>"index", "sort"=>"sample"}`

### Actual behavior

The output from params is `{"controller"=>"test", "action"=>"index", ""=>":sampl", "sort"=>"e"}`

### System configuration
**Rails version**: 8.0.0.rc1

**Ruby version**: 3.3.5

**Comment by ruyrocha** [comment]
By checking the specs, it seems the exception is raised only when you define routes using `resources`:

https://github.com/rails/rails/blob/e9fa6faae66781b71a9fade02234927d3b8e4b94/actionpack/test/dispatch/routing_test.rb#L3487-L3495

If you remove one colon from the parameter like this

```ruby
get 'sort:sort', to: 'test#index'
```

Then you can reference it in your controller like `params[:sort]`, e.g:

```ruby
(rdbg) params
#<ActionController::Parameters {"controller"=>"test", "action"=>"index", "sort"=>":sample"} permitted: false>
(ruby) params[:sort]
":sample"
(ruby) params[:sort].delete_prefix(':')
"sample"
(rdbg) 
```

**Comment by paratiger** [comment]
We have systems that run using URLs such as `articles/sort:title` and `articles/page:2` for pagination.

Up until Rails 8.0.0rc1 the routes `articles/sort::sort` and `articles/page::page` correctly pass the parameters:

```
articles/sort:title => {controller: 'articles', action: 'index', sort: 'title'}
articles/page:2 => {controller: 'articles', action: 'index', page: '2'}
```

What has changed that no longer allows these routes with double '::' from working?

Note that these are not `resource` routes

**Comment by rafaelfranca** [comment]
Can you check if https://github.com/rails/rails/pull/53453 fixes it?

**Comment by paratiger** [comment]
@rafaelfranca I can confirm that #53453 fixes the issue (test case and live sites)

Thanks for looking into this

**Comment by paratiger** [comment]
Closing this as fixed with #53453
